{
    "id": "SOLR-7451",
    "title": "\u201dNot enough nodes to handle the request\u201c when inserting data to solrcloud",
    "details": {
        "components": [
            "clients - java"
        ],
        "type": "Bug",
        "labels": "",
        "fix_versions": [],
        "affect_versions": "5.1",
        "status": "Resolved",
        "resolution": "Not A Problem",
        "priority": "Major"
    },
    "description": "I use solr5.1.0 and deploy one node with solrcloud\uff0cand create a collection with 1 shard and 2 replica\uff0cwhen i use solrj to insert data\uff0cit throw \u201dNot enough nodes to handle the request\u201c\uff0cbut if i create collection with 1 shard and 1 replica\uff0cit can insert successfully\uff0calso i create another replica with admin api\uff0cit still work fine\uff0cno longer throw that exception\nthe full exception stack\nException in thread \"main\" org.apache.solr.client.solrj.SolrServerException: org.apache.solr.common.SolrException: Not enough nodes to handle the request\n\tat org.apache.solr.client.solrj.impl.CloudSolrClient.requestWithRetryOnStaleState(CloudSolrClient.java:929)\n\tat org.apache.solr.client.solrj.impl.CloudSolrClient.requestWithRetryOnStaleState(CloudSolrClient.java:922)\n\tat org.apache.solr.client.solrj.impl.CloudSolrClient.requestWithRetryOnStaleState(CloudSolrClient.java:922)\n\tat org.apache.solr.client.solrj.impl.CloudSolrClient.requestWithRetryOnStaleState(CloudSolrClient.java:922)\n\tat org.apache.solr.client.solrj.impl.CloudSolrClient.requestWithRetryOnStaleState(CloudSolrClient.java:922)\n\tat org.apache.solr.client.solrj.impl.CloudSolrClient.requestWithRetryOnStaleState(CloudSolrClient.java:922)\n\tat org.apache.solr.client.solrj.impl.CloudSolrClient.request(CloudSolrClient.java:782)\n\tat org.apache.solr.client.solrj.SolrRequest.process(SolrRequest.java:135)\n\tat org.apache.solr.client.solrj.SolrClient.add(SolrClient.java:107)\n\tat org.apache.solr.client.solrj.SolrClient.add(SolrClient.java:72)\nCaused by: org.apache.solr.common.SolrException: Not enough nodes to handle the request\n\tat org.apache.solr.client.solrj.impl.CloudSolrClient.sendRequest(CloudSolrClient.java:1052)\n\tat org.apache.solr.client.solrj.impl.CloudSolrClient.requestWithRetryOnStaleState(CloudSolrClient.java:839)\n\t... 10 more",
    "attachments": {},
    "issue_links": {},
    "comments": [
        {
            "date": "2015-06-24T10:42:07+0000",
            "author": "jatin sutaria",
            "content": "Is there any update on this? I am facing the same issue with 1 shard and 3 replicas for a collection.? I am using Solr 5.2 ",
            "id": "comment-14599231"
        },
        {
            "date": "2015-08-05T12:54:21+0000",
            "author": "Markus Jelsma",
            "content": "We have seen this too, on Solr 5.2, 1 shard, three replica's.  ",
            "id": "comment-14655302"
        },
        {
            "date": "2015-08-12T16:26:02+0000",
            "author": "Guido",
            "content": "Hi, I am experiencing the same issue on Solr 5.2.1, 1 collection with 4 shards without any replica. ",
            "id": "comment-14693756"
        },
        {
            "date": "2015-08-12T17:41:06+0000",
            "author": "Ishan Chattopadhyaya",
            "content": "I tried to reproduce this with 5.2.1,\n\nbin/solr -c\nbin/solr create -c demo -shards 1 -replicationFactor 3\ncurl \"http://localhost:8983/solr/demo/update?commit=true\" -H 'Content-type:application/json' -d '[{\"id\" : \"MyTestDocument\", \"title_t\" : \"This is just a test\"}]'\ncurl \"http://localhost:8983/solr/demo/select?q=*:*&wt=json&omitHeader=true\"\n\n\nThis worked for me. I'll try to use the Collections API to create the collection next to see if I can reproduce. ",
            "id": "comment-14693909"
        },
        {
            "date": "2015-08-12T18:24:42+0000",
            "author": "Erick Erickson",
            "content": "Could you add your cluster state for the collection when it fails? It's in your admin UI/cloud/tree either under\n\n/clusterstate.json (older style)\nor\n/collections/collection_name/state.json (newer 5x versions).\n\nOh, and to help Ishan, it would also be good if you added the exact commands you used to create the collection in the failing case and the code you use for adding docs. Otherwise it's really hard for anyone to have confidence that they're doing the same thing you're doing.\n\nBest,\nErick\n ",
            "id": "comment-14693972"
        },
        {
            "date": "2015-08-12T19:09:13+0000",
            "author": "Guido",
            "content": "Hello guys, thanks for your prompt attention on it. Actually, while trying to troubleshoot the problem I ended up manually modifying the state.json file of my collection. Basically, I changed the 'state' from 'down' to 'active', then I restarted Solr. It started without any error on the log file and I was able to query my collection again. I am still trying to work on it so if I see any other error I will post a new comment here, but so far the manual change on the json file seems to solve the problem. ",
            "id": "comment-14694024"
        },
        {
            "date": "2015-08-12T19:33:55+0000",
            "author": "Erick Erickson",
            "content": "Uhhh, this is like, really, really dangerous. But it seems like the answer to your original problem was accurately reported: there were no replicas active for a given shard.\n\nThere remains the question of why there were no replicas, the Solr logs would help there. It's rather doubtful that the problem was with the collection creation, although I can't rule that out.\n\nIshan Chattopadhyaya maybe close this ticket then? ",
            "id": "comment-14694053"
        },
        {
            "date": "2015-08-12T20:00:40+0000",
            "author": "Guido",
            "content": "Hello. I am not sure why this is so dangerous (surely you know it better than me). I would be happy if you want to elaborate it. Anyway, I believe that the problem is strictly related to a plugin inside the /lib/ directory: as a second test, I tried to modify my solrconfig to avoid the use of the custom plugin and I was able to create the collection. Then, I deployed the plugin inside the 'lib' directory, modified the solrconfig and reloaded the collection. In this way I did not get the problem. I hope that this helps you: I always get problems while trying to create a collection which uses a custom plugin and this 2-steps always helps me. ",
            "id": "comment-14694081"
        },
        {
            "date": "2015-08-12T21:26:12+0000",
            "author": "Erick Erickson",
            "content": "Well, Solr tries very hard to not lose updates. Part of the processing is to be sure that the replica that becomes the leader has all the updates possible is contained in the leader election process. If a shard can't elect a leader, then it's in a non-deterministic state as far as being consistent. By going in and manually adjusting the state of a node to active, you're bypassing all that processing.\n\nIt's not necessarily bad, it's just that there are risks.\n\nBTW, it would have been really helpful to tell us you had a custom plugin in the first place. Time spent trying to reproduce a problem isn't well spent without complete information.\n\nI'll close this ticket. ",
            "id": "comment-14694200"
        },
        {
            "date": "2015-08-12T21:26:27+0000",
            "author": "Erick Erickson",
            "content": "Apparently a problem with a custom plugin. ",
            "id": "comment-14694202"
        },
        {
            "date": "2015-08-12T21:51:39+0000",
            "author": "Guido",
            "content": "Thanks, really sorry if I have not originally specified the presence of a custom plugin but I did not realize it was related to the problem, since I had issues with the custom plugins in the past but I never experienced this particular issue. I am not sure that the 'custom plugin' is the root cause of the problem also for the other 3 users which experienced this issue, but thank you for your prompt help and support. ",
            "id": "comment-14694234"
        },
        {
            "date": "2015-08-13T03:51:03+0000",
            "author": "laigood",
            "content": "Thanks all,when i remove my custom analyzer and try again,it work fine!just curious why custom analyzer can case this. ",
            "id": "comment-14694653"
        },
        {
            "date": "2015-08-13T09:38:36+0000",
            "author": "Guido",
            "content": "Hello laigood, in my particular case the problem was not the custom plugin itself but the deployment of it that failed while creating the collection. In fact, if I create the collection and subsequently I deploy the same exact plugin, I get no issues. ",
            "id": "comment-14694958"
        },
        {
            "date": "2015-08-13T15:23:23+0000",
            "author": "Erick Erickson",
            "content": "Guido:\n\nRight, Creating a collection pushes the solrconfig file out and as each core starts it looks for the plugin. Now, this should work fine if you have a lib directive in your solrconfig.xml file the points to your custom jar file. You could even put your custom jar file in the same place as other the Solr jars, but I prefer a distinct directory.\n\nSo the process should be\n1> deploy Solr\n2> modify your solrconfig.xml file to contain a <lib..../> directive somewhere on each machine. Push it to ZK.\n3> copy your custom jar to the \"somewhere\" in <2>.\n4> create collections.\n\nIf that doesn't work, we need to know. I'd start by using an absolute path in my solrconfig since it's simplest, then move to relative if I really needed to.\n\nFWIW,\nErick ",
            "id": "comment-14695352"
        },
        {
            "date": "2015-08-20T08:25:20+0000",
            "author": "Rakesh",
            "content": "if i create collection with 1 shard and 1 replica, I am getting the same error at random interval. Once I encounter this error, Nothing works till i restart my Zookeeper and Solr servers.. ",
            "id": "comment-14704501"
        },
        {
            "date": "2015-08-20T09:05:51+0000",
            "author": "Guido",
            "content": "Hi Erick,\n\nUnfortunately I am extremely busy in these days, but as soon as possible I will give a try with the custom jar and I will let you know.\n\nKind Regards,\n\nGuido ",
            "id": "comment-14704548"
        },
        {
            "date": "2015-08-20T15:18:52+0000",
            "author": "Erick Erickson",
            "content": "Guido:\n\nOK, \n\nRakesh:\n\nPlease start these kinds of discussions on the user's list, see: http://lucene.apache.org/solr/resources.html. When you do post there, include the log file for the downed node. We try to reserve JIRAs for known code issues, not usage problems. I suspect that if you look at your log files for the various Solr nodes (NOT the indexing program) you'll see something more enlightening.\n ",
            "id": "comment-14705102"
        },
        {
            "date": "2015-08-25T06:31:53+0000",
            "author": "Markus Jelsma",
            "content": "We just saw this issue again on 5.2.0. Repeated Solr or Zookeeper restarts 'fixed' the issue. I believe it is a client/zookeeper issue, there were no INFO nor WARNING notifications in the server log. ",
            "id": "comment-14710687"
        },
        {
            "date": "2015-08-25T08:01:41+0000",
            "author": "Ishan Chattopadhyaya",
            "content": "Markus Jelsma, can you please share a ZK dump during such a situation? ",
            "id": "comment-14710831"
        },
        {
            "date": "2017-06-20T09:26:39+0000",
            "author": "Dana Sava",
            "content": "We are using Solr 5.2.1, with one shard, 2 replicas for each collection. We have several collections (around 70). When posting data to Solr (using solj) on multiple threads, on multiple collections in the same time, we receive frequently this error. We believe it might be a big or if it's not, than can someone detail under which conditions this exception is thrown? ",
            "id": "comment-16055436"
        },
        {
            "date": "2017-06-20T09:28:44+0000",
            "author": "Dana Sava",
            "content": "I forgot to mention, when the exception is logged, the Solr Cloud is looking very good - all the nodes are up (green). ",
            "id": "comment-16055438"
        }
    ]
}